### PR TITLE
BUG: fix a bug where AngleFormatterLocator and ScalarFormatterLocator wouldn't respect matplotlib.rc's axes.unicode_minus

### LINKED ISSUE #15898: Plotting with wcsaxes does not respect matplotlib's unicode_minus
### Description

The tick formatter for `wcsaxes` does not respect matplotlib's `axes.unicode_minus` setting, producing ticks with a `U+002d HYPHEN-MINUS` (-) sign instead of the visually more appealing and correct `U+2212 MINUS SIGN` (−).

### Expected behavior

Respect the `axes.unicode_minus` setting (which is the default) and produce typographically correct minus signs.

A simple fix is to run the strings produced by `formatter()` through [`matplotlib.ticker.Formatter.fix_minus()`](https://matplotlib.org/stable/api/ticker_api.html#matplotlib.ticker.Formatter.fix_minus).

### How to Reproduce

Plot any `wcsaxes` plot with negative ticks.

### Versions

macOS-10.16-x86_64-i386-64bit
Python 3.10.13 (main, Sep 11 2023, 08:39:02) [Clang 14.0.6 ]
astropy 6.0.0
Numpy 1.26.0
pyerfa 2.0.1.1
Scipy 1.11.3
Matplotlib 3.7.2

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
Fixes #15898

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/visualization/wcsaxes/formatter_locator.py
Comment (ntessore) [inline astropy/visualization/wcsaxes/formatter_locator.py:65]: In this case, maybe it is better to just implement the logic here?
```suggestion
    if not rcParams['axes.unicode_minus']:
        return labels
    minus = "\N{MINUS SIGN}"
    return [minus + s[1:] if s.startswith("-") else s for s in labels]
```

File: astropy/visualization/wcsaxes/tests/test_formatter_locator.py
Comment (ntessore) [inline astropy/visualization/wcsaxes/tests/test_formatter_locator.py:468]: Could just use "\N{MINUS SIGN}" here

**Review by ntessore — COMMENTED** [review]
Thanks for the prompt fix @neutrinoceros! I just left two comments to take or leave

File: astropy/visualization/wcsaxes/tests/test_formatter_locator.py
Comment (neutrinoceros) [reply astropy/visualization/wcsaxes/tests/test_formatter_locator.py:468]: ow, I didn't know about that ! I'll see what other reviewers think about it first, as I'm not sure how obscure this feature is and I'm worried it wouldn't be as readable to maintainers.

File: astropy/visualization/wcsaxes/tests/test_display_world_coordinates.py
Comment (neutrinoceros) [inline astropy/visualization/wcsaxes/tests/test_display_world_coordinates.py:34]: Here I think using unicode strings as suggested would make both the change itself and hypothetical test failures clearer, so I'll change it.

**Comment by neutrinoceros** [comment]
All failing image comparison tests look good to me but this is new territory to me so before I push new hashes, can I ask @pllim for confirmation ? (I just wonder wether I should wait on reviews to update hashes or not)

**Comment by ntessore** [comment]
There's a complication: [`simplify_labels()`](https://github.com/astropy/astropy/blob/21a7ed197c57adb69385e403e7751632099fd92d/astropy/visualization/wcsaxes/ticklabels.py#L123) strips repeated strings from labels, and does not ignore the unicode minus sign.

**Comment by neutrinoceros** [comment]
@ntessore nice catch. I can fix this too. Can you suggest a way to demonstrate the remaining bug ?

**Comment by ntessore** [comment]
Simple reproducer (uses `ScalarFormatterLocator`):

```py
import numpy as np
import matplotlib.pyplot as plt

from astropy.wcs import WCS

wcs = WCS({
    "CRVAL1": 0.,
    "CRVAL2": 0.,
    "CRPIX1": 1000,
    "CRPIX2": 1000,
    "CD1_1": 0.05,
    "CD1_2": 0.,
    "CD2_1": 0.,
    "CD2_2": 0.05,
})

plt.subplot(projection=wcs)
plt.imshow(np.random.rand(2000, 2000))
plt.show()
```

Without unicode minus:
before

With unicode minus:
after

**Comment by pllim** [comment]
If you think the new outputs are okay, I think you can ship the new image hashes with your PR to make CI pass? I am not very familiar with this still but it should be documented at https://docs.astropy.org/en/latest/development/testguide.html#image-tests-with-pytest-mpl 

Ultimately, visualization maintainers make the call: @larrybradley @Cadair @astrofrog

**Comment by neutrinoceros** [comment]
Yeah I read the docs but it doesn't mention anything about reviewers' input. Oh well, hashes are cheap to push anyway.

**Comment by neutrinoceros** [comment]
@ntessore just for reference I think the following patch would probably fix the problem you pointed to regarder `simplify_labels`, but we'd need a test (not necessarily a complete real world scenario) to demonstrate that the current behaviour isn't satisfactory.
```patch
diff --git a/astropy/visualization/wcsaxes/ticklabels.py b/astropy/visualization/wcsaxes/ticklabels.py
index 5b37f0ed05..aeb955a899 100644
--- a/astropy/visualization/wcsaxes/ticklabels.py
+++ b/astropy/visualization/wcsaxes/ticklabels.py
@@ -125,6 +125,8 @@ class TickLabels(Text):
         Figure out which parts of labels can be dropped to avoid repetition.
         """
         self.sort()
+        skippable_chars = "0123456789."
+        skippable_chars += '\N{MINUS SIGN}' if rcParams['axes.unicode_minus'] else '-'
         for axis in self.world:
             t1 = self.text[axis][0]
             for i in range(1, len(self.world[axis])):
@@ -140,7 +142,7 @@ class TickLabels(Text):
                 for j in range(len(t1) - 1):
                     if t1[j] != t2[j]:
                         break
-                    if t1[j] not in "-0123456789.":
+                    if t1[j] not in skippable_chars:
                         start = j + 1
                 t1 = self.text[axis][i]
                 if start != 0:
 ```

**Comment by ntessore** [comment]
@neutrinoceros I agree this would fix the issue, which does appear on your branch for me.

```py
import os
import numpy as np
import matplotlib.pyplot as plt

from astropy.wcs import WCS

wcs = WCS({
    "CRVAL1": 0.,
    "CRVAL2": 0.,
    "CRPIX1": 1000,
    "CRPIX2": 1000,
    "CD1_1": 0.05,
    "CD1_2": 0.,
    "CD2_1": 0.,
    "CD2_2": 0.05,
})

plt.subplot(projection=wcs)
plt.imshow(np.random.rand(2000, 2000))
plt.title("git commit " + os.popen("git rev-parse --short=7 HEAD").read())
plt.show()
```

image

**Comment by neutrinoceros** [comment]
Oh, I see now: the problem is that the minus sign was dropped from the "-40" on the y axis and from the "-20" on the x axis ! thank you for your vigilance !

**Comment by neutrinoceros** [comment]
I pushed the patch as is, hoping that some image comparison test will highlight the difference, which would allow not to have a dedicated test.

**Comment by neutrinoceros** [comment]
I think tests should all pass now but I note that there's something I don't yet get with pytest-mpl being used in hybrid mode (hashes + baseline images), which results in some tests "passing" globally only because hashes take precedent over image baseline, but that doesn't seem to be addressed in our docs. I'll dig into `pytest-mpl`'s for how to work with hybrid mode and if I learn anything useful I might open a separate PR here to improve our documentation.

Screenshot 2024-01-18 at 11 54 09

**Comment by neutrinoceros** [comment]
Actually it is already addressed in our docs: https://docs.astropy.org/en/latest/development/testguide.html#generating-reference-images (just not _immediately_ where I most expected it)

**Comment by pllim** [comment]
> just not immediately where I most expected it

I can't find it half the time either. If you have idea to make it more obvious, please propose PR. Thanks!

**Comment by neutrinoceros** [comment]
Kindly pinging @Cadair @astrofrog @larrybradley for review (no rush !) 😄

File: astropy/visualization/wcsaxes/formatter_locator.py
Comment (larrybradley) [reply astropy/visualization/wcsaxes/formatter_locator.py:65]: I assume `matplotlib.ticker.Formatter.fix_minus` checks the `rcParams`?

Comment (neutrinoceros) [reply astropy/visualization/wcsaxes/formatter_locator.py:65]: It does ! see https://github.com/matplotlib/matplotlib/blob/fd5f1640cc1f3b5676d3242e10bd069e18d75f6c/lib/matplotlib/ticker.py#L246-L260

Comment (larrybradley) [reply astropy/visualization/wcsaxes/formatter_locator.py:65]: Thanks!

**Review by larrybradley — APPROVED** [review]
LGTM, thanks!

**Comment by pllim** [comment]
Giles webhook giving me problem, so I cannot see the artifacts.

**Comment by pllim** [comment]
Actually... https://app.circleci.com/pipelines/github/astropy/astropy/19231/workflows/d1979c6d-ebce-46e8-9992-ddb27775d496/jobs/119379/artifacts says "After 30 days, all artifacts associated with a build automatically expire. Since this run is over 30 days old, the artifacts may have been deleted."

Are you able to rebase and trigger a new build? Thanks!

**Comment by pllim** [comment]
I don't grok this stuff but a quick glance looks fine: https://output.circle-artifacts.com/output/job/818d0993-c804-4559-b716-48922e344060/artifacts/0/results/fig_comparison.html?q=test_custom_frame

Since Larry approved, I merged this. Thanks!

**Pull request merged**
